returnn/torch/data/pipeline.py

```diff
@@ -309,6 +311,99 @@ def __iter__(self):
+            any segments make sure your largest bucket allows segments larger than your longest training segment.
+        :param length_key: data key to take as length measure
+        :param _kwargs: unused args
+        """
+        self._dataset = dataset
+        self._length_key = length_key
+
+        assert buckets, "empty bucket batching configuration"
+        if not all(size > 0 and max_seqs > 0 for size, max_seqs in buckets):
+            raise ValueError(f"bucket sizes and max seqs in bucket must be positive")
+        self._max_seq_lens, self._max_bucket_sizes = zip(*sorted(buckets, key=lambda b: b[0]))
+        assert len(set(self._max_seq_lens)) == len(self._max_seq_lens), "seq len boundaries must all be unique"
+
+    def __iter__(self):
+        """:return: generator applying bucket ordering on the data"""
+        buckets: List[List[Dict[str, numpy.ndarray]]] = [[] for _ in range(len(self._max_seq_lens))]
+
+        for data_dict in self._dataset:
+            data_dict: Dict[str, numpy.ndarray]
+            data_size = data_dict[self._length_key].shape[0]
+            bucket_idx = bisect.bisect_left(self._max_seq_lens, data_size)
+            if bucket_idx >= len(self._max_seq_lens):
+                # seg is too long, drop it
+                continue
+            buckets[bucket_idx].append(data_dict)
+            if len(buckets[bucket_idx]) >= self._max_bucket_sizes[bucket_idx]:
+                yield buckets[bucket_idx]
+                buckets[bucket_idx] = []
+
+        non_empty_buckets = [b for b in buckets if b]
+        yield from non_empty_buckets
+
+
+def init_batching(
+    dataset: torch.utils.data.IterableDataset, config: Config, train: bool
+) -> torch.utils.data.IterableDataset:
+    """
+    Batches the segments in the given dataset given the config settings.
+
+    :param dataset: dataset whose segments to group into batches
+    :param config: RETURNN config
+    :param train: whether we are in training or in inference mode
+    """
+    custom_batching = config.typed_value("custom_batching", None)
+    if custom_batching is None:
+        batch_size = config.typed_value("batch_size", -1)
+        batch_size = config.typed_value(f"batch_size_{'train' if train else 'dev'}", batch_size)
+        assert batch_size != -1, f"batch_size or batch_size_{'train' if train else 'dev'} not defined in config"
+        max_seqs = config.typed_value("max_seqs", -1)
+        batches_dataset = BatchingIterDataPipe(dataset, batch_size=batch_size, max_seqs=max_seqs)
+        return batches_dataset
+
+    assert isinstance(custom_batching, dict)
+    assert "class" in custom_batching
+    custom_batching = custom_batching.copy()
+    clazz = custom_batching.pop("class")
+    if isinstance(clazz, str):
+        batchers = {c.__name__: c for c in [BatchingIterDataPipe, BucketOrderingIterDataPipe]}
+        make_batcher = batchers[clazz]
+    elif isinstance(clazz, type):
+        make_batcher = clazz
```

Comment:
```suggestion
        pass
```
(Maybe move the `assert issubclass(cls, torch.utils.data.IterDataPipe)` that I suggested after this `if`.)

Reply:
> Maybe move the `assert issubclass(cls, torch.utils.data.IterDataPipe)` that I suggested after this `if`.

Maybe we can also just let it be any `callable`, not just subclass of `IterDataPipe`. In case it is some generic function, you can also pass `train` and `**get_fwd_compat_kwargs()` as kwargs to this callable.

And then after it was called, check:
```python
assert isinstance(batches_dataset, torch.utils.data.IterDataPipe)
```